lib/datadog/statsd/sender.rb

```diff
@@ -32,13 +54,31 @@ def rendez_vous
       def add(message)
         raise ArgumentError, 'Start sender first' unless message_queue
 
+        # if the thread does not exist, we assume we are running in a forked process,
+        # empty the message queue and message buffers (these messages belong to
+        # the parent process) and spawn a new companion thread.
+        if !sender_thread.alive?
+          @mx.synchronize {
+            # a call from another thread has already re-created
+            # the companion thread before this one acquired the lock
+            break if sender_thread.alive?
+            @logger.debug { "Statsd: companion thread is dead, re-creating one" } if @logger
+
+            message_queue.close if CLOSEABLE_QUEUES
+            @message_queue = nil
+            message_buffer.reset
+            message_buffer.reset_telemetry
+            start
+          }
+        end
+
         message_queue << message
       end
```

Comment:
🤔  Hmm I still see a potential issue here, similar to the ones in `#flush` and `#rendez_vous`: with some poor timing of `stop`, by the time we get to the `message_queue << message` line, `message_queue` may be `nil`.

(And its cousin issue, `@sender_thread` being `nil`)

I think part of the issue is that we have the background thread setting these two things to `nil` without ever synchronizing with any other threads, which can get surprised by this at many points in their execution. While we could expand the synchronization even more, that seems to me to be a bit heavy-handed, especially since we have a thread-safe construct (`Queue`) that we're building around in this class.

Here's my suggestion:
1. Construct the `Queue` in `#initialize`
2. Never set it to `nil` or close it (but we may `#clear` it when restarting the background thread or after a stop). This enables us to always know that we can safely use it and call methods on it.
3. Only synchronize when mutating `@sender_thread` -> starting it, changing it (when it dies), or setting it to `nil` (when it finishes due to `#stop`). Reading `@sender_thread` for checks is OK to do without locks.

This is just a suggestion, so feel free to ignore and do something else. But I think this class as it is, is hiding a lot of complexity introduced by the shared mutable state, truly the more I look the more I see potential issues.

Let me know if you'd like to pair on this; perhaps that way we can get this across the finish line without so much async back-and-forth and rework.

Reply:
That is why at first I've used the mutex to synchronize the close as well, it would avoid timing with the `#stop` call since the `#stop` would be able to run only if it has the lock. I'll see what change represents your suggestion, and I agree that this class ships a lot of complexity now...